### DISCUSSION TITLE
How to filter the NT8PropertyGrid Control view

### DISCUSSION THREAD

**Discussion by MicroTrendsTom** [MicroTrendsLtd/NinjaTrader8#64] [Q&A] [answered]
Assistance required:
"NinjaTrader Property Grid" , how to filter its view by a group of property attributes within its SelectedObject.
There is no issues with adding tabs or elements in tabs - it is the "property grid" behaviors or lack of which poses the challenge etc - and no open source either etc.

For the sample below i had to create data classes for each view shown in the tab and thus the property grid selected object was the data class instance:
![image](https://user-images.githubusercontent.com/24366913/156975771-44bcd419-a82c-4a36-8fea-a8f4315cd652.png)

The current implementation shown here uses 1 view and is harder to navigate at a glance
![image](https://user-images.githubusercontent.com/24366913/156975813-45ddaf86-3617-455c-8e21-cf707372f89e.png)

So it would be very easy to use a tab control and add an instance of the property grid to each tab and set thePropertyGrid.SelectedObject as the strategy this so it displays all the properties in the derived class also - but it would show all public properties in the SelectedObject

The Question is:

without building out a set of classes 1 per view 
1A)How to filter the view to a set of property's by attribute
1B) Provide databound functionality 

Maybe Selectedobject= this.GetPropsWhere "AttributeABC,AttributeTradeSetup") etc.

**Comment by jeronymite** [comment]
An alternative approach may be to use the "tab" buttons to cause PropertyGrid Categories to to be collapsed/expanded as appropriate. Very simple to implement, and whilst not as visually clean as hiding the unwanted categories, the result still provides a "focused by category" view. So, for each "tab" button, collapse all categories not covered by that tab and expand all the categories covered by it. Just a thought.

**Reply by MicroTrendsTom** [reply]
Yup nice idea any code snippets welcome :-) 
to Hide and show a specific set of groups by some attribute/group would be a great win also

**Comment by jeronymite** [comment, marked as answer]
The _PropertyGrid_ has arrays of Categories and Properties. For example, for a declared _PropertyGrid_ "_pgrid_", use _pgrid.Categories_ and _pgrid.Properties_. Each of the Categories and Properties in those arrays has a boolean property of _IsExpanded_, which can be read or set.

Having identified the Categories and Properties that correspond with individual tabs/buttons, simply cycle through all Categories and Properties and set _IsExpanded_ to true for the ones you want to focus on and false for those you do not want to focus on. As you probably already know, you can create Categories for your Properties using _[Display(GroupName = AnyGroupNameYouWant)]_ in your _PropertyGrid_ class.

You may also find the code for the _PropertyGrid_ developed by Denys Vuika, the one used by NinjaTrader 8, to be of use:  (https://github.com/DenysVuika/WPG) Not sure if the github version is identical to the one in NT8, but it certainly formed the basis of it. You may even find something there to help you do the hide/show you're after. ... or even support for tabs ;-)

**Reply by MicroTrendsTom** [reply]
Yes of course an excellent idea - iterative get and set of the underlying groups within the WPF Elements - show/hide expand according to the criteria etc

I think this will certainly be the route for now to take as the other idea with manipulating properties and classes was not looking too good
So unless that way is found we later we will set this as the answer and go from here

many thanks for the input!

**Comment by jeronymite** [comment]
Most welcome. One last comment, do take a look at this (https://github.com/DenysVuika/WPG/wiki/Display-Property-Categories-in-Separate-Tabs) to see if it may suit your purposes.

**Reply by MicroTrendsTom** [reply]
Yup its almost there  a nice feature tabbed view for a group.
What we need is a group of groups per tab 
So One way or another  it would be possible to iterate through the WPF elements and hide show inline
or perhaps move elements  to a tab etc
will have to post some ideas and experiments later

Ty once again :-)
